src/dataregistry/registrar.py

```diff
@@ -86,91 +205,60 @@ def register_dataset(self, name, relative_path, version_major,
+            return None
 
-        # Confirm new dataset exists
+        # deal with version string
+        special = version in ['major', 'minor', 'patch']
+        if not special:
+            v_fields = parse_version_string(version)
+            version_string = version
+        else:
+            # Generate new version fields based on previous entries
+            # with the same name field and same suffix
+            v_fields = bump_version(name, version, version_suffix,
+                                    dataset_table, self._engine)
+            version_string = '.'.join(str(v_fields.values()))
+
+        # Get dataset characteristics; copy if requested
         if not is_dummy:
-            dest =  form_dataset_path(self._owner_type, self._owner,
-                                      relative_path, self._root_dir)
-            if old_location:
-                loc = old_location
-            else:
-                loc = dest
-
-            if os.path.isfile(loc):
-                dataset_organization = "file"
-            elif os.path.isdir(loc):
-                dataset_organization = "directory"
-            else:
-                raise FileNotFoundError(f"Dataset {loc} not found")
-
-            # Get metadata on dataset.
-            if verbose:
-                tic = time.time()
-                print("Collecting metadata...", end="")
-            if dataset_organization == "directory":
-                num_files, total_size = get_directory_info(loc)
-            else:
-                num_files = 1
-                total_size = os.path.getsize(loc)
-            if verbose:
-                print(f"took {time.time()-tic:.2f}s")
-
-            if old_location:
-                # copy to dest.  For directory do recursive copy
-                # for now always copy; don't try to handle sym link
-                # Assuming we don't want to copy any metadata (e.g.
-                # permissions)
-                if verbose:
-                    tic = time.time()
-                    print(f"Copying {num_files} files ({total_size/1024/1024:.2f} Mb)...",end="")
-                if dataset_organization == "file":
-                    # Create any intervening directories
-                    os.makedirs(os.path.dirname(dest), exist_ok=True)
-                    copyfile(old_location, dest)
-                elif dataset_organization == "directory":
-                    copytree(old_location, dest, copy_function=copyfile)
-                if verbose:
-                    print(f"took {time.time()-tic:.2f}")
+            dataset_organizaation, num_files, total_size = \
+                self._handle_data(relative_path, old_location)
         else:
-            # In the case of a dummy dataset
             dataset_organization = "dummy"
             num_files = 0
             total_size = 0
 
+        # If no execution_id is supplied, create a minimal entry
```

Comment:
Thinking ahead for this. If the user registers a bunch of datasets and forgets about the execution, then wants to create an execution later and associate those datasets with it post registry. Then this would leave a bunch of left over entries in the execution table.

Unless we:

- Dont allow post-entry changes, but i think people will often forget and need to change things post-entry
- Allow them to overwrite certain fields of the execution table entry when its update post-entry, rather than replace the execution associated with a dataset with a new entry (which i think is ok?)

The idea behind an auto entry with every dataset is so there is something to link dependencies ?

Reply:
I would vote for the second option: (someday) providing the ability to update certain columns in `execution`. 
And yes, the point was just to have something there. It's meant for the case where the user either doesn't know much about how the dataset was created or doesn't care.      Then, in case they do care about dependencies, we should provide a way to add them in later, something which I think you've mentioned before.